### PR TITLE
Add getter_functions in modules, functions scope 

### LINKED ISSUE #4371: help: linker error when linking compiled binaries to executable
After working around bugs, it seems we can now finally incorporate the subroutine `translv` through a module into the file `snap_main.f90`. In the compilation phase, we find that the file takes an unusually long time to compile. Later in the linking phase of the process, we get an error stating 'linker failure'.

GFortran compiles and executes successfully with the changes.

Branch: https://github.com/lanl/SNAP/compare/main...Pranavchiku:SNAP:lf4

### Linker error
```console
.
.
.
/usr/bin/ld: snap_main.o:LFortran:(.text+0x3cafb): more undefined references to `__lcompilers_get_nang' follow
/usr/bin/ld: snap_main.o: in function `__module_translv_module_translv':
LFortran:(.text+0x40c9b): undefined reference to `__lcompilers_get_ng1'
/usr/bin/ld: LFortran:(.text+0x40cd4): undefined reference to `__lcompilers_get_ng2'
/usr/bin/ld: LFortran:(.text+0x40cdb): undefined reference to `__lcompilers_get_nthreads'
/usr/bin/ld: LFortran:(.text+0x40cf6): undefined reference to `__lcompilers_get_ng'
/usr/bin/ld: LFortran:(.text+0x41996): undefined reference to `__lcompilers_get_ng1'
/usr/bin/ld: LFortran:(.text+0x419b3): undefined reference to `__lcompilers_get_ng1'
/usr/bin/ld: LFortran:(.text+0x41a4b): undefined reference to `__lcompilers_get_ng2'
/usr/bin/ld: LFortran:(.text+0x41a53): undefined reference to `__lcompilers_get_nthreads'
/usr/bin/ld: LFortran:(.text+0x42d0f): undefined reference to `__lcompilers_get_ng'
/usr/bin/ld: LFortran:(.text+0x43186): undefined reference to `__lcompilers_get_ng1'
/usr/bin/ld: LFortran:(.text+0x431a3): undefined reference to `__lcompilers_get_ng1'
/usr/bin/ld: LFortran:(.text+0x435db): undefined reference to `__lcompilers_get_ng'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
The command 'clang -o gsnap global.o utils.o version.o plib.o geom.o sn.o data.o control.o input.o setup.o dealloc.o translv.o solvar.o outer.o expxs.o inner.o sweep.o octsweep.o dim1_sweep.o dim3_sweep.o output.o time.o mms.o analyze.o thrd_comm.o mkba_sweep.o snap_main.o  -L"/home/saurabh-kumar/Projects/System/lfortran/src/bin/../runtime" -Wl,-rpath,"/home/saurabh-kumar/Projects/System/lfortran/src/bin/../runtime" -llfortran_runtime -lm' failed.
make: *** [Makefile:131: gsnap] Error 10
```

### LINKED ISSUE #4435: bug : Getter functions doesn't get imported from mod files
related to #4371 
**MRE**
```fortran 
module c
    integer ::j=20
end module c

module a
    use c
    contains
    subroutine fdf()
        integer :: o(j,2)
        print *,j
        print *, size(o)
    end subroutine
end module a

program g1
    use a
    call fdf()
end program g1

```
This code compiles normally with no errors.
The error occurs when we move `program g1` into another file, the getter function `_lcompilers_get_ j` doesn't get imported into the ASR.

### PR DESCRIPTION
**Pull request by assem2002**
fixes #4435  
fixes #4371

Instead of adding getter_functions into the global scope of the ASR (which makes it get lost after deserializing the dependencies back), We now add them in the module scope or the nesting function.

- We couldn't just leave the function in the global scope because (as I know) we don't serialize standalone function that aren't contained in modules.
***
- A use case that could have an issue is : compiling an `.f90` file that has only standalone functions and trying to link it with other files (The getter_functions would get added into the global scope at this case).

### CODE REVIEW

**Review by Pranavchiku — COMMENTED** [review]
Can you check if the following example works?

```fortran
module xx
integer :: yy = 12
end module

subroutine sub(x)
use xx
integer, dimension(yy) :: x
print *, x
```

File: integration_tests/arrays_50_mod.f90 [hunk 1]
```diff
@@ -0,0 +1,24 @@
+
+
```
Comment (Pranavchiku) [inline integration_tests/arrays_50_mod.f90:2]:
```suggestion
```

File: integration_tests/arrays_50_mod.f90 [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        if (j /= 20) error stop
+        if (j2 /= 30) error stop
+        print *, size(o)    
+        if (size(o) /= 600) error stop
+    end subroutine fdf
+end module arrays_50_mod_a
+
+
+ 
+
```
Comment (Pranavchiku) [inline integration_tests/arrays_50_mod.f90:24]:
```suggestion
```

**Review by Pranavchiku — APPROVED** [review]
This looks good to me. The only reason to do it in global scope was to avoid creation of such access functions but if there is problem in serialising them ( which I am not sure of ) then this approach is fine. Also, in latest main we do not have any provision to not create `_lcompilers_get_xx` function if already present. So I think this is fine. Great work!

**Comment by assem2002** [comment]
I think we have a similar test for the example you provided.

**Comment by Pranavchiku** [comment]
With this PR, does the linker work?

**Comment by assem2002** [comment]
Thank you. 
The linker works, the test in this PR ensures that.

**Comment by Pranavchiku** [comment]
I enabled auto merge so that you guys can make progress on SNAP. I think this PR is good.

**Comment by assem2002** [comment]
Thanks you :smile:

**Comment by assem2002** [comment]
Do you think the global function would've worked out if we used flags `--implicit-interface` and `--generate-object-code` ?

**Pull request merged**